src/java/org/apache/cassandra/service/pager/AggregationQueryPager.java

```diff
@@ -326,7 +354,9 @@ protected QueryPager updatePagerLimit(QueryPager pager,
          */
         private final PartitionIterator fetchSubPage(PageSize subPageSize)
         {
-            return consistency != null ? subPager.fetchPage(subPageSize, consistency, queryState, queryStartNanoTime)
+            checkTimeout();
+
+            return consistency != null ? subPager.fetchPage(subPageSize, consistency, queryState, System.nanoTime())
```

Comment:
Shouldn't we send `queryStartNanoTime`? Why override it?

Reply:
That's the crux of the separate timeout for aggregation queries.
Currently, because of sending `queryStartNanoTime`, a page read will end with a timeout if it crosses the range read timeout counted from the beginning of the aggregation query (`queryStartNanoTime`). But we want to time out the aggregation query only when it crosses the, usually much larger, aggregation query timeout.

In other words, before the change, the whole aggregation query must fit within the range read timeout.
After the change, each single page read must fit within the range read timeout, but the overall aggregation query is subject to the aggregation query timeout (checked separately).

Reply:
I find it different from an existing logic so I got confused. Currently `Monitorable` is monitoring query execution and checking if we've reach the timeout. One idea could be to re-use that logic and maybe override timeout in `PartitionRangeReadCommand` for aggregate queries?

Reply:
https://github.com/datastax/cassandra/blob/main/src/java/org/apache/cassandra/db/PartitionRangeReadCommand.java#L256 could be easily tweaked to return aggregate query timeout?